test-app/src/main/java/com/ustadmobile/meshrabiya/testapp/viewmodel/NearbyTestViewModel.kt

```diff
@@ -0,0 +1,273 @@
+package com.ustadmobile.meshrabiya.testapp.viewmodel
+
+import android.util.Log
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.meshrabiya.lib_nearby.nearby.NearbyVirtualNetwork
+import com.ustadmobile.meshrabiya.log.MNetLogger
+import com.ustadmobile.meshrabiya.testapp.appstate.AppUiState
+import com.ustadmobile.meshrabiya.testapp.appstate.FabState
+import com.ustadmobile.meshrabiya.testapp.server.ChatMessage
+import com.ustadmobile.meshrabiya.testapp.server.ChatServer
+import com.ustadmobile.meshrabiya.vnet.AndroidVirtualNode
+import com.ustadmobile.meshrabiya.vnet.VirtualPacket
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.update
+import kotlinx.coroutines.launch
+import org.kodein.di.DI
+import org.kodein.di.instance
+import java.net.InetAddress
+import java.nio.ByteBuffer
+import kotlin.random.Random
+
+class NearbyTestViewModel(
+    di: DI
+) : ViewModel() {
+    private val _uiState = MutableStateFlow(NearbyTestUiState())
+    val uiState: StateFlow<NearbyTestUiState> = _uiState.asStateFlow()
+
+    private val virtualNode: AndroidVirtualNode by di.instance()
+    private var nearbyNetwork: NearbyVirtualNetwork? = null
+    private var chatServer: ChatServer? = null
+    private var isNetworkInitialized = false
+
+    private val logger = object : MNetLogger() {
+        override fun invoke(priority: Int, message: String, exception: Exception?) {
+            val logMessage = "${MNetLogger.priorityLabel(priority)}: $message"
+            viewModelScope.launch {
+                _uiState.update { currentState ->
+                    currentState.copy(logs = currentState.logs + logMessage)
+                }
+            }
+            if (exception != null) {
+                Log.e(TAG_NEARBY_TEST, message, exception)
+            }
+        }
+
+        override fun invoke(priority: Int, message: () -> String, exception: Exception?) {
+            invoke(priority, message(), exception)
+        }
+    }
+
+    init {
+        initializeNearbyNetwork()
+        updateAppUiState()
+    }
+
+    private fun initializeNearbyNetwork() {
+        try {
+            nearbyNetwork = NearbyVirtualNetwork(
```

Comment:
This **should not** be done here. Initializing the virtual network every time a viewmodel starts is obviously wrong.